packages/flutter/lib/src/widgets/router.dart

```diff
@@ -512,23 +510,29 @@ class _RouterState<T> extends State<Router<T>> with RestorationMixin {
     _routeInformationReportingTaskScheduled = false;
 
     if (_routeInformation.value != null) {
-      final RouteInformation routeInformation = _routeInformation.value!;
+      final RouteInformation oldRouteInformation = widget.routeInformationProvider!.value;
+      final RouteInformation currentRouteInformation = _routeInformation.value!;
       switch (_currentIntentionToReport) {
         case _IntentionToReportRouteInformation.none:
           assert(false, '_reportRouteInformation must not be called with _IntentionToReportRouteInformation.none');
           return;
         case _IntentionToReportRouteInformation.ignore:
+          if (oldRouteInformation.location != currentRouteInformation.location ||
+              oldRouteInformation.state != currentRouteInformation.state) {
+            widget.routeInformationProvider!.routerUpdatesRouteInformation(currentRouteInformation);
+          }
           break;
         case _IntentionToReportRouteInformation.maybe:
-          if (_lastSeenLocation != routeInformation.location) {
-            widget.routeInformationProvider!.routerReportsNewRouteInformation(routeInformation);
+          if (oldRouteInformation.location != currentRouteInformation.location) {
```

Comment:
It's odd that the Router makes the decision whether to replace or append by calling different API on the routeInformationProvider since this cannot be customized. Should the Router maybe tell the RouteInformationProvider that there's some new RouteInformation and then it's up to the RouteInformationProvider to decide how to message that to the engine (i.e. whether to replace or append the history entry)?

Reply:
we will need to change the API of routerReportsNewRouteInformation to know whether the Router want to force push a new route information or ignore it. I feel it may be more confusing than right now.

Reply:
I still find it strange that we bake it into the router that a state-only change replaces the current history entry. What if I want to use the back button of my browser to go to a previous state of my single-page app that doesn't expose different URLs?

Is replacing really the only use case we have to cover?

Reply:
(Maybe adding the `navigate` and `neglect` methods to the router was a mistake and the router should always provide the current RouteInformation to the InformationProvider which can then decide whether a new entry should be created?)

Reply:
`navigate` was for the case where user want to create history entry for state only change.

Do you mean providing the API in InformationProvider to choose whether user wants to create entry or replace entry? It can replace the API, but I feel it may be harder to use the API, for example, if the developer wants a state change causes by a button press to create a history entry, the informationProvider needs to have some way to figure out a the of state change.